examples/5_secret_storage.py

```diff
@@ -0,0 +1,419 @@
+        return res
+
+    async def delete(
+        self,
+        team: str,
+        namespace: str,
+        entry_key: str,
+        revision: Union[int, None] = None,
+    ) -> keybase1.KVDeleteEntryResult:
+        h = await self.hmac_key(team, namespace, entry_key)
+        res = await self.kvstore.delete(team, namespace, h, revision)
+        res.entry_key = entry_key
+        return res
+
+    async def get(
+        self, team: str, namespace: str, entry_key: str
+    ) -> keybase1.KVGetResult:
+        h = await self.hmac_key(team, namespace, entry_key)
+        res = await self.kvstore.get(team, namespace, h)
+        res.entry_key = entry_key
+        return res
+
+    async def list_entrykeys(
+        self, team: str, namespace: str
+    ) -> keybase1.KVListEntryResult:
+        res = await self.kvstore.list_entrykeys(team, namespace)
+        if res.entry_keys:
+            for e in res.entry_keys:
+                if not e.entry_key.startswith("_"):
+                    get_res = await self.kvstore.get(team, namespace, e.entry_key)
+                    e.entry_key = json.loads(get_res.entry_value)[self.KEY_KEY]
+        return res
+
+
+class RentalClient:
+    """
+        Wraps a CachedKVStoreClient to expose methods to handle tool rentals.
+    """
+
+    NAMESPACE = "rental"
+
+    def __init__(self, kvstore: CachedKVStoreClient):
+        self.kvstore = kvstore
+
+    async def lookup(self, team, tool) -> keybase1.KVGetResult:
+        res = await self.kvstore.get(team, self.NAMESPACE, tool)
+        return res
+
+    async def add(
+        self, team, tool
+    ) -> Union[keybase1.KVPutResult, keybase1.KVGetResult]:
+        info: Dict[str, str] = {}
+        expected_revision = 1
+        cached = self.kvstore.get_cached(tool)
+        if cached is not None:
+            # if tool already exists, propagate existing info
+            if cached["info"]:
+                info = cached["info"] if type(info) is dict else {}
+            expected_revision = cached["revision"] + 1
+        res = await self.kvstore.put(
+            team, self.NAMESPACE, tool, info, expected_revision
+        )
+        return res
+
+    async def remove(
+        self, team, tool
+    ) -> Union[keybase1.KVDeleteEntryResult, keybase1.KVGetResult, None]:
+        expected_revision = 1
+        cached = self.kvstore.get_cached(tool)
+        if cached is not None:
+            expected_revision = cached["revision"] + 1
+        res = await self.kvstore.delete(team, self.NAMESPACE, tool, expected_revision)
+        return res
+
+    async def update_reservation(
+        self, team, user, tool, day, reserve=True
+    ) -> Union[keybase1.KVPutResult, keybase1.KVGetResult]:
+        # note: if you reserve or unreserve a not-added or deleted tool, it will add the tool
+        info: Dict[str, str] = {}
+        expected_revision = 1
+        cached = self.kvstore.get_cached(tool)
+        if cached is not None:
+            if cached["info"]:
```

Comment:
yeah - it's sub-optimal that the RentalClient knows about how the CachedClient arranges its internal data structures